File: Question/5_Longest_Palindromic_Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int start = 0;
        int len = 0;
        for(int i = 0; i < s.size(); ++i) {
            palindrome(s, i, i, start, len);
            palindrome(s, i, i + 1, start, len);
        }
        return s.substr(start, len);
    }
    //find the palindrome substring from left and right
    void palindrome(string& s, int left, int right, int& start, int& len) {
        while(left >= 0 && right < s.size() && s[left] == s[right]) {
            --left;
            ++right;
        }
        if(len < right - left - 1) {
            len = right - left - 1;
            start = left + 1;
        }
    }
};
//dp(i, j) [i, j] whether a palindrome stirng
//dp(i, j) = {1 | i == j}
//         = {s[i] == s[j] | j == i + 1}
//         = {s[i] == s[j] && dp(i + 1, j - 1)}
class Solution {
public:
    string longestPalindrome(string s) {
        if(s.size() == 0 || s.size() == 1)
            return s;
        int start = 0;
        int len = 0;
        vector<vector<int>> dp(s.size(), vector<int>(s.size(), 0));
        //int dp[s.size()][s.size()];
        for(int j = 0; j < s.size(); ++j) {
            dp[j][j] = 1;
            for(int i = 0; i <= j; ++i) {
                dp[i][j] = (s[i] == s[j] && (j - i < 2 || dp[i + 1][j - 1]));
                if(dp[i][j] && len < j - i + 1) {
                    start = i;
                    len = j - i + 1;
                }
            }
        }

        return s.substr(start, len);
    }
};

File: Question_Go/5_Longest_Palindromic_Substring.go
//dp(i, j) [i, j] whether a palindrome stirng
//dp(i, j) = {1 | i == j}
//         = {s[i] == s[j] | j == i + 1}
//         = {s[i] == s[j] && dp(i + 1, j - 1)}

//go的二维数组初始化很麻烦！！！
func longestPalindrome(s string) string {
    dp := make([][]bool, len(s))
    start, length := 0, 0

    for j := 0; j < len(s); j++ {
        dp[j] = make([]bool, len(s))
        dp[j][j] = true
        for i := 0; i <= j; i++ {
            dp[i][j] = (s[i] == s[j] && (j - i < 2 || dp[i + 1][j - 1]));
            if dp[i][j] && length < j - i + 1 {
                start = i;
                length = j - i + 1;
            }
        } 
    }

    return s[start : start + length]
}